exposure-notification/src/main/java/fi/thl/covid19/exposurenotification/eu/FederationGatewayService.java

```diff
@@ -0,0 +1,114 @@
+package fi.thl.covid19.exposurenotification.eu;
+
+import com.google.protobuf.ByteString;
+import com.google.protobuf.InvalidProtocolBufferException;
+import fi.thl.covid19.exposurenotification.batch.BatchFileService;
+import fi.thl.covid19.exposurenotification.diagnosiskey.DiagnosisKeyDao;
+import fi.thl.covid19.exposurenotification.diagnosiskey.IntervalNumber;
+import fi.thl.covid19.exposurenotification.diagnosiskey.v1.TemporaryExposureKey;
+import fi.thl.covid19.exposurenotification.error.EfgsOperationException;
+import fi.thl.covid19.proto.EfgsProto;
+import org.springframework.stereotype.Service;
+
+import java.time.Instant;
+import java.time.LocalDate;
+import java.time.format.DateTimeFormatter;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+@Service
+public class FederationGatewayService {
+
+    private final FederationGatewayClient client;
+    private final BatchFileService batchFileService;
+    private final DiagnosisKeyDao dd;
+
+    public FederationGatewayService(
+            FederationGatewayClient client,
+            BatchFileService batchFileService,
+            DiagnosisKeyDao diagnosisKeyDao
+    ) {
+        this.client = client;
+        this.batchFileService = batchFileService;
+        this.dd = diagnosisKeyDao;
+    }
+
+    public void doOutbound() {
+        long operationId = dd.startOutboundOperation();
+        try {
+            List<TemporaryExposureKey> localKeys = dd.fetchAvailableKeysForEfgs(operationId);
+            handleOutbound(transform(localKeys), operationId);
+            dd.finishOperation(operationId);
+        } catch (Throwable t) {
```

Comment:
Don't catch throwable. Use finally-block for "whatever happens" operation closing. 

For instance, you could use a variable outside the block to mark success and then set operation end-result in a finally block with something like if(operationFinished) dd.finish... etc. Or maybe a parametrized function rather than 2 separate ones. Finally block also avoids the need to catch and rethrow anything as the original exception keeps on going.

Reply:
See above.